nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/iot/GetAWSIoT.java

```diff
@@ -0,0 +1,153 @@
+    "The processor keeps open a WebSocket connection and will automatically renew the " +
+    "connection to overcome Amazon's service limit on maximum connection duration. Depending on " +
+    "your set up QoS the processor will miss some messages (QoS=0) or receives messages twice (QoS=1) " +
+    "while reconnecting to AWS IoT WebSocket endpoint. We strongly recommend you to make use of " +
+    "processor isolation as concurrent subscriptions to an MQTT topic result in multiple message receiptions.")
+@SeeAlso({ GetAWSIoTShadow.class })
+@WritesAttributes({
+        @WritesAttribute(attribute = "aws.iot.mqtt.endpoint", description = "AWS endpoint this message was received from."),
+        @WritesAttribute(attribute = "aws.iot.mqtt.topic", description = "MQTT topic this message was received from."),
+        @WritesAttribute(attribute = "aws.iot.mqtt.client", description = "MQTT client which received the message."),
+        @WritesAttribute(attribute = "aws.iot.mqtt.qos", description = "Underlying MQTT quality-of-service.")
+})
+public class GetAWSIoT extends AbstractAWSIoTProcessor {
+
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(
+                    PROP_QOS,
+                    PROP_TOPIC,
+                    PROP_ENDPOINT,
+                    PROP_KEEPALIVE,
+                    PROP_CLIENT,
+                    AWS_CREDENTIALS_PROVIDER_SERVICE,
+                    REGION));
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return Collections.singleton(REL_SUCCESS);
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        // init to build up mqtt connection over web-sockets
+        init(context);
+        if (mqttClient != null && mqttClient.isConnected()) {
+            try {
+                // subscribe to topic with configured qos in order to start receiving messages
+                mqttClient.subscribe(awsTopic, awsQos);
+            } catch (MqttException e) {
+                getLogger().error("Error while subscribing to topic " + awsTopic + " with client-id " + mqttClient.getClientId() + " caused by " + e.getMessage());
+            }
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final List messageList = new LinkedList();
+        // check if connection is about to terminate
+        if (isConnectionAboutToExpire()) {
+            MqttWebSocketAsyncClient _mqttClient = null;
+            try {
+                // before subscribing to the topic with new connection first unsubscribe
+                // old connection from same topic if subscription is set to QoS 0
+                if (awsQos == 0) mqttClient.unsubscribe(awsTopic);
```

Comment:
The client is not unsubscribed first when using QoS 1 but is when using QoS 0. Are there any benefits to unsubscribing first?